scripts/format_code.py

```diff
@@ -0,0 +1,227 @@
+     because -noformat_file was set.
+   2: The application wasn't configured properly and could not execute.
+"""
+import io
+import os
+import subprocess
+import sys
+
+from absl import app
+from absl import flags
+
+# Flag Definitions:
+FLAGS = flags.FLAGS
+
+flags.DEFINE_boolean("git_diff", False, "Use git-diff to assemble a file list")
+flags.DEFINE_string("git_range", "main..", "the range string when using "
+  "git-diff.")
+flags.DEFINE_multi_string("f", None, "Append the filename to the list of "
+  "files to check.")
+flags.DEFINE_multi_string("d", None, 
+  "Append the directory to the file list to format.")
+flags.DEFINE_boolean("r", False, 
+  "Recurse through the directory's children When formatting a target directory.")
+flags.DEFINE_boolean("format_file", True, "Format files in place.")
+flags.DEFINE_boolean("verbose", False, "Execute in verbose mode.")
+
+# Constants:
+FILE_TYPE_EXTENSIONS = (".cpp", ".cc", ".c", ".h")
+"""Tuple: The file types to run clang-format on.
+Used to filter out results when searching across directories or git diffs.
+"""
+
+
+# Functions:    
+def does_file_need_formatting(filename):
+  """Executes clang-format on the file to determine if it includes any
+  formatting changes the user needs to fix.
+  Args:
+    filename (string): path to the file to check.
+  
+  Returns:
+    bool: True if the file requires format changes, False if formatting would produce
+    an identical file.
+  """
+  args = ['clang-format', '-style=file', '-output-replacements-xml', filename]
+  result = subprocess.run(args, stdout=subprocess.PIPE)
+  for line in result.stdout.decode('utf-8').splitlines():
+    if line.strip().startswith("<replacement "):
+      return True
+  return False    
+
+def format_file(filename):
+  """Executes clang-format on the file to reformat it according to our
+  project's standards.  If -format_in_place is defined, the file will be
+  formatted in place, with the results overwriting the previous version
+  of the file. Otherwise the result of the file reformat will be logged
+  to standard out.
+
+  Args:
+   filename (string): path to the file to format.
+  """
+  args = ['clang-format', '-style=file', '-i', filename]
+  subprocess.run(args)
+
+def git_diff_list_files():
+  """Compares the current branch to master to assemble a list of source
+  files which have been altered.
+  
+  Returns:
+    A list of file paths for each file in the git diff list with an extension
+    matching one of those in FILE_TYPE_EXTENSIONS.
+  """
+  # The ACMRT filter lists files that have been (A)dded (C)opied (M)odified
+  # (R)enamed or (T)ype changed.
+  args = ['git', 'diff', FLAGS.git_range, '--name-only', '--diff-filter=ACMRT']
+  filenames = []
+  if FLAGS.verbose:
+    print()
+    print('Executing git diff to detect changed files, git_range: "{0}"'
+        .format(FLAGS.git_range))
+  result = subprocess.run(args, stdout=subprocess.PIPE)
```

Comment:
Worth testing if the output is captured without capture_output. Also, I think subprocess.PIPE is not required. This might work,
`subprocess.run(args, capture_output=True)`

Reply:
'capture_output' was added in 3.7, which perhaps not all of our users will have installed, so I'm going with this option.